### PR TITLE
feat: implement sessions using an encrypted cookie

### PR DESCRIPTION
**Pull request by definitelynobody**
I fixed an issue with jobs not getting cleaned up properly so please read my changes carefully to `jobs.rs` and main.rs

Blocked by #123

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -532,42 +541,41 @@ async fn root_post(
...
+        if lock.i2j.remove(&uuid).is_some() {
             debug!("timeout for: {}", uuid);
-            let mut lock = user.write().await;
-            if lock.data.job().as_ref().map(|j| j.uuid) == Some(uuid) {
-                lock.data.kill_job().await;
-            }
+        }
+
+        if lock.u2i.get(&user) == Some(&uuid) {
+            let _ = lock.u2i.remove(&user);
```
Comment (definitelynobody) [inline src/main.rs:573]: using `let _ = lock.u2i.remove(&user);` and not killing the job will mean ports and docker containers will not be freed. the job will not be killed and we will run out of resources

**Comment by rjzak** [comment]
Seems to need a rebase, some of the changes look like they're from #123

**Comment by definitelynobody** [comment]
@rjzak should be okay now

**Review by rjzak — APPROVED** [review]
Looks good to me

**Pull request merged**